### PR TITLE
pump/: Fix get L0 files num and trigger vlog.gcTS step by step

### PR DESCRIPTION
**Pull request by july2993**
### What problem does this PR solve? 
Refine gc of pump

### What is changed and how it works?
1, Add a API to trigger gc
2, Check L0 files by *GetProperty* instead of Stats
Stats can't get the right L0 files num, see:
https://github.com/syndtr/goleveldb/pull/283/files
3, call vlog.gcTS step by step, help free space quickly
4,   update goleveldb dependency
    only one new commit
    https://github.com/syndtr/goleveldb/commit/02440ea7a28525b3079783ad9c27083994a42366
    Fix the stat about levels of the Stats API

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
call `debug/gc/trigger`

Code changes

Side effects

 Related changes

 - Need to cherry-pick to the release branch
 - Need to be included in the release note

### CODE REVIEW

File: pump/server.go [hunk 1]
```diff
@@ -500,26 +503,29 @@ func (s *Server) gcBinlogFile() {
 		case <-s.ctx.Done():
 			log.Info("gcBinlogFile exit")
 			return
+		case <-s.triggerGC:
```
Comment (GregoryIan) [inline pump/server.go]: it's better to use output time from `triggerGC` to  do gc

Comment (GregoryIan) [reply pump/server.go]: and it's better to add these code in pr #646

File: pump/storage/storage.go [hunk 2]
```diff
@@ -610,15 +610,22 @@ func (a *Append) doGCTS(ts int64) {
 	}
 
 	for {
-		var stats leveldb.DBStats
-		err := a.metadata.Stats(&stats)
+		nStr, err := a.metadata.GetProperty("leveldb.num-files-at-level0")
 		if err != nil {
-			log.Error("Stats failed", zap.Error(err))
+			log.Error("GetProperty failed", zap.Error(err))
```
Comment (GregoryIan) [inline pump/storage/storage.go]:
```suggestion
			log.Error("get  `leveldb.num-files-at-level0` property failed", zap.Error(err))
```

File: pump/storage/storage.go [hunk 3]
```diff
@@ -610,15 +610,22 @@ func (a *Append) doGCTS(ts int64) {
...
 			time.Sleep(5 * time.Second)
 			continue
 		}
-		if len(stats.LevelTablesCounts) > 0 && stats.LevelTablesCounts[0] >= l0Trigger {
-			log.Info("wait some time to gc cause too many L0 file", zap.Int("files", stats.LevelTablesCounts[0]))
+
+		l0Num, err := strconv.Atoi(nStr)
+		if err != nil {
+			log.Error("parse int failed", zap.String("str", nStr), zap.Error(err))
```
Comment (GregoryIan) [inline pump/storage/storage.go]:
```suggestion
			log.Error("parse `leveldb.num-files-at-level0` result to int  failed", zap.String("str", nStr), zap.Error(err))
```

File: pump/storage/storage.go [hunk 4]
```diff
@@ -610,15 +610,22 @@ func (a *Append) doGCTS(ts int64) {
...
 
 	for {
-		var stats leveldb.DBStats
-		err := a.metadata.Stats(&stats)
+		nStr, err := a.metadata.GetProperty("leveldb.num-files-at-level0")
 		if err != nil {
-			log.Error("Stats failed", zap.Error(err))
+			log.Error("GetProperty failed", zap.Error(err))
 			time.Sleep(5 * time.Second)
 			continue
```
Comment (GregoryIan) [inline pump/storage/storage.go]: return directly?

File: pump/storage/storage.go [hunk 5]
```diff
@@ -654,6 +665,10 @@ func (a *Append) doGCTS(ts int64) {
 			}
 			break
 		}
+
+		if len(lastKey) > 0 {
```
Comment (GregoryIan) [inline pump/storage/storage.go]: good job!

**Comment by suzaku** [comment]
LGTM

File: pump/storage/storage.go [hunk 6]
```diff
@@ -654,6 +663,10 @@ func (a *Append) doGCTS(ts int64) {
 			}
 			break
 		}
+
```
Comment (GregoryIan) [inline pump/storage/storage.go]: add log for kv gc and vlog gc

**Comment by GregoryIan** [comment]
Rest LGTM

File: pump/storage/storage.go (hunk 6 above)
Comment (july2993) [reply pump/storage/storage.go]: add in https://github.com/pingcap/tidb-binlog/pull/648/commits/5e91e874a6db935d100e92c87c540e58663a0a39

**Comment by GregoryIan** [comment]
LGTM

**Pull request merged**